Refactor omniComments.js #5436

JosephineKwa opened the pull request:
Fixes #5436

JosephineKwa commented:
Hey @junhaoyap , for your review as well :)

damithc commented on src/main/webapp/js/omniComment.js:
```diff
@@ -217,18 +218,16 @@ $(document).ready(function() {
-                commentListRegionForFeedbackResponse.prev().prev().show();
-                commentListRegionForFeedbackResponse.prev().prev().prev().show();
+        if (commentToShow.prop('class').toString().includes(classNameForCommentsInFeedbackResponse)) {
+            if (commentToShow.parent().find('li[style*="display: none"]').length !== commentToShow.parent().find('li').length) {
```
Can use explaining variables to clarify the meaning of these boolean conditions?
Also, looks like you can `and`  the two conditions (because there are no `else`) to reduce nesting.

junhaoyap replied:
Since we're already modifying this let's take this chance to make the check more understandable :s

damithc commented on src/main/webapp/js/omniComment.js:
```diff
@@ -256,21 +255,19 @@ $(document).ready(function() {
-                commentListRegionForFeedbackResponse.prev().prev().hide();
-                commentListRegionForFeedbackResponse.prev().prev().prev().hide();
+        if (commentToHide.prop('class').toString().includes(classNameForCommentsInFeedbackResponse)) {
+            if (commentToHide.parent().find('li[style*="display: none"]').length === commentToHide.parent().find('li').length) {
```
These two lines look similar to the previous case. If they are, may be extract as function?

junhaoyap replied:
Almost, one of them checks of equality and one checks for inequality, although after looking at it for some time I still can't wrap my head around what it means :\

junhaoyap commented on src/main/webapp/js/omniComment.js:
```diff
@@ -217,18 +218,16 @@ $(document).ready(function() {
-                commentListRegionForFeedbackResponse.prev().show();
-                commentListRegionForFeedbackResponse.prev().prev().show();
-                commentListRegionForFeedbackResponse.prev().prev().prev().show();
+        if (commentToShow.prop('class').toString().includes(classNameForCommentsInFeedbackResponse)) {
```
We can use jQuery's ```.hasClass()``` instead, it's more semantical

junhaoyap commented on src/main/webapp/js/omniComment.js:
```diff
@@ -245,9 +244,9 @@ $(document).ready(function() {
-        if ($(comment).prop('class').toString().includes(classNameForCommentsInStudentRecords)) {
-            var studentCommentPanel = $(comment).closest('.student-comments-panel');
-            var studentCommentPanelBody = $(comment).parent();
+        if (commentToShow.prop('class').toString().includes(classNameForCommentsInStudentRecords)) {
```
Ditto

junhaoyap commented on src/main/webapp/js/omniComment.js:
```diff
@@ -256,21 +255,19 @@ $(document).ready(function() {
-                commentListRegionForFeedbackResponse.prev().hide();
-                commentListRegionForFeedbackResponse.prev().prev().hide();
-                commentListRegionForFeedbackResponse.prev().prev().prev().hide();
+        if (commentToHide.prop('class').toString().includes(classNameForCommentsInFeedbackResponse)) {
```
Same here!

junhaoyap commented on src/main/webapp/js/omniComment.js:
```diff
@@ -287,9 +284,9 @@ $(document).ready(function() {
-        if ($(comment).prop('class').toString().includes(classNameForCommentsInStudentRecords)) {
-            var studentCommentPanel = $(comment).closest('.student-comments-panel');
-            var studentCommentPanelBody = $(comment).parent();
+        if (commentToHide.prop('class').toString().includes(classNameForCommentsInStudentRecords)) {
```
Ditto

junhaoyap commented on src/main/webapp/js/omniComment.js:
```diff
@@ -256,21 +255,19 @@ $(document).ready(function() {
+        if (commentToHide.prop('class').toString().includes(classNameForCommentsInFeedbackResponse)) {
+            if (commentToHide.parent().find('li[style*="display: none"]').length === commentToHide.parent().find('li').length) {
+                var commentListRegionForFeedbackResponse = commentToHide.closest('tr');
+                // a fbResponse in instructorCommentsPage (html) is made up of 4 rows in a table
```
Comment seems redundant now that we are using ```.siblings().andSelf()``` which is semantical by itself

junhaoyap commented on src/main/webapp/js/omniComment.js:
```diff
@@ -217,18 +218,16 @@ $(document).ready(function() {
+        if (commentToShow.prop('class').toString().includes(classNameForCommentsInFeedbackResponse)) {
+            if (commentToShow.parent().find('li[style*="display: none"]').length !== commentToShow.parent().find('li').length) {
+                var commentListRegionForFeedbackResponse = commentToShow.closest('tr');
+                // a fbResponse in instructorCommentsPage (html) is made up of 4 rows in a table
```
Same :D

junhaoyap commented on src/main/webapp/js/omniComment.js:
```diff
@@ -7,9 +7,10 @@ $(document).ready(function() {
-        var start = $(location).attr('href').indexOf('#');
-        var end = $(location).attr('href').length;
-        var rowId = $(location).attr('href').substring(start, end);
+        var url = $(location).attr('href');
```
This can simply be ```location.href```, location is a JavaScript object that already has the href property in and of itself, wrapping it as a jQuery object and accessing by ```.attr``` is redundant

junhaoyap commented:
@JosephineKwa looks good! Just a few minor repeated improvements that we can do and one major one since we can't really understand what the check is checking for

JosephineKwa commented:
Changes made:
1. Extracted long boolean conditions into explaining variables
2. Used hasClass rather than checking the string of the element's class property
3. Removed unnecessary comments
4. Replaced `$(location)` with `location.href`
5. Simplified showCommentAndItsPanel by removing unnecessary conditions. This is because if the comment is shown, the panels will definitely have to be shown. 

I wasn't able to combine the two functions as one of them is used for showing elements and the other is used for hiding elements.

In addition I found a bug which causes the feedback question panel not to be hidden when all the feedback responses are hidden, and a similar bug which causes the feedback session panel not to be hidden when all the feedback question panels are hidden. I will create an issue for that.

junhaoyap commented on src/main/webapp/js/omniComment.js:
```diff
@@ -1,15 +1,13 @@
-    
     function isRedirectToSpecificComment() {
-        return $(location).attr('href').indexOf('#') !== -1;
+        return location.href.indexOf('#') !== -1;
```
I remember including a polyfill of https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/includes, try to use it if possible, it's semantically more meaningful than ```.indexOf...```

wkurniawan07 replied:
@junhaoyap yes you did, because I had to `eslint-disable` that part of the code 😛

junhaoyap replied:
@wkurniawan07 ohh what happened to that part of the code?

wkurniawan07 replied:
Nothing, just violating [this rule](http://eslint.org/docs/rules/no-extend-native).

junhaoyap commented:
I see the new updates have some APIs like ```.addBack()``` that I am not familiar with yet so give me more time to review this ok? :) Going sleep now as it's about 1:10 am here :p

JosephineKwa commented:
Edit: I also realised using ```siblings().andSelf()``` was wrong as it hid/showed more than what was necessary. I ended up using:
```
commentListRegionForFeedbackResponse.prevUntil('.feedback-response-giver-recipient-row')
        .addBack().prev('.feedback-response-giver-recipient-row').addBack().show();
```
But it doesn't really look like much of an improvement over the multiple `prev()`s :frowning:

junhaoyap commented on src/main/webapp/js/omniComment.js:
```diff
@@ -217,84 +215,80 @@ $(document).ready(function() {
+            // The following statement shows the rows between comment list region row and giver recipient row
+            // including the comment list region row and giver recipient row.
+            // See stackoverflow post: http://stackoverflow.com/questions/2770588/jquery-prevuntil-include-start-selector-and-end-selecter
+            commentListRegionForFeedbackResponse.prevUntil('.feedback-response-giver-recipient-row')
```
Hmm, yeah it seems this way of doing it isn't that much of an improvement.. will it be possible to generate a class for all the rows that should be hidden and just select the class and hide all of them instead? Would it be too much work? If yes perhaps we can shift it to a new PR :)

junhaoyap commented on src/main/webapp/js/omniComment.js:
```diff
@@ -217,84 +215,80 @@ $(document).ready(function() {
+                // The following statement hides the rows between comment list region row and giver recipient row
+                // including the comment list region row and giver recipient row.
+                // See stackoverflow post: http://stackoverflow.com/questions/2770588/jquery-prevuntil-include-start-selector-and-end-selecter
+                commentListRegionForFeedbackResponse.prevUntil('.feedback-response-giver-recipient-row')
```
Ditto

junhaoyap commented:
Otherwise, looking better than before, I'll let you decide whether we should improve that part in this PR or in another one! I will be away on a road trip for about 3 days so I won't be able to review until then :)

JosephineKwa commented:
Ready for review again! I've replaced the indexOf with includes and generated classes for the table rows to be hidden/shown

wkurniawan07 commented:
@junhaoyap let's review this soon?

junhaoyap commented:
Indeed - I have forgotten, thanks for the ping @wkurniawan07, will look through all my other PRs as well :)

junhaoyap commented on src/main/webapp/js/omniComment.js:
```diff
@@ -224,13 +224,9 @@ $(document).ready(function() {
-            commentListRegionForFeedbackResponse.prevUntil('.feedback-response-giver-recipient-row')
-                    .addBack().prev('.feedback-response-giver-recipient-row').addBack().show();
+            var rowsToShowClassName = commentListRegionForFeedbackResponse.attr('class');
+            $('.' + rowsToShowClassName).show();
```
Unrelated to this, in ES6 we can do string templating which will make this more obvious, maybe one day we might be able to make TEAMMATES ES6 ready :)

junhaoyap commented:
Looks great, there could be more improvements but they'll be outside the scope of this PR, apologies for taking so long, slipped my mind after coming back from the trip :)

junhaoyap commented:
@damithc @JosephineKwa please take a look!

wkurniawan07 commented:
@JosephineKwa let's merge this soon?

JosephineKwa commented:
I decided to revert the latest commit `replace indexOf with includes` as `includes` doesn't seem to work on my firefox browser

wkurniawan07 commented:
@JosephineKwa how did you discover that it didn't work? Is it by calling the function from a random website?

JosephineKwa commented:
Yup pretty much :open_mouth:

wkurniawan07 commented:
That explains. That function is polyfilled via `common.js` so you can't try it from just about anywhere.

wkurniawan07 commented:
@JosephineKwa do you want to undo the revert and merge this one (after getting green) now? I still can wait for this to be merged before concluding this week's milestone.

JosephineKwa commented:
Yup I will do that now!
